cmd/agent/daemon/enrichment/enrichers.go

```diff
@@ -0,0 +1,149 @@
+package enrichment
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"path/filepath"
+	"strconv"
+	"syscall"
+
+	castpb "github.com/castai/kvisor/api/v1/runtime"
+	"github.com/castai/kvisor/pkg/containers"
+	"github.com/castai/kvisor/pkg/logging"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/minio/sha256-simd"
+)
+
+type fileHashCacheKey string
+type ContainerForCgroupGetter func(cgroup uint64) (*containers.Container, bool, error)
+
+func EnrichWithFileHash(log *logging.Logger, containerForCgroupGetter ContainerForCgroupGetter, procFS fs.StatFS) EventEnricher {
+	cache, err := lru.New[fileHashCacheKey, []byte](1024)
+	if err != nil {
+		// This case can never happen, as err is only thrown if cache size is <0, which it isn't.
+		panic(err)
+	}
+
+	return &fileHashEnricher{
+		log:                      log,
+		containerForCgroupGetter: containerForCgroupGetter,
+		procFS:                   procFS,
+		cache:                    cache,
+	}
+}
+
+type fileHashEnricher struct {
+	log                      *logging.Logger
+	containerForCgroupGetter ContainerForCgroupGetter
+	procFS                   fs.StatFS
+	cache                    *lru.Cache[fileHashCacheKey, []byte]
+}
+
+func (enricher *fileHashEnricher) Enrich(ctx context.Context, e *castpb.Event) {
+	exec := e.GetExec()
+	if exec == nil || exec.Path == "" {
+		return
+	}
+
+	cont, found, err := enricher.containerForCgroupGetter(e.CgroupId)
```

Comment:
Inside tracer_decode we already get container and full event context. Would be nice if we can reuse already fetched data for enrichments. Maybe it then makes sense to run enrichment inside ebpf tracer as this is it outputs final castpb.Event anyway.

Reply:
Good point, then I could also pass along the event context. Overall it should not be too bad though, as the function we pass in as the cgroup getter only returns cached values from the map.